File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    max_len := 0
    freq := map[byte]int{}
    j := 0
    for i := range s {
        char := s[i]
        freq[char]++

        win_len := i - j + 1

        for j < i && win_len > len(freq) {
            freq[s[j]]--

            if freq[s[j]] == 0 { delete(freq, s[j]) }

            j++
            win_len = i - j + 1
        }


        if win_len <= len(freq) {
            max_len = max(max_len, win_len)
        }
    }

    return max_len
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> st;
        int j = 0;
        int maxlen = 0;
        for(int i=0; i<s.length(); i++){
            char x = s[i];
            if (st.count(x)){
                int len = i - j;
                maxlen = max(len, maxlen);
                while(j < i && st.count(x)){
                    st.erase(s[j]);
                    j++;
                }
            }
            st.insert(x);
        }
        maxlen = max(maxlen, int(st.size()));
        return maxlen;
    }
};